### DISCUSSION TITLE
Best way to check for duplicates of one column grouped by another?

### DISCUSSION THREAD

**Discussion by Aariq** [rstudio/pointblank#394] [Q&A] [answered]
What's the best way to check, for example, that ID numbers are not duplicated within years.   I suspect there is a better way than what I'm trying to do:
``` r
library(tidyverse)
```

    #> Warning: package 'tidyr' was built under R version 4.0.5

``` r
df <- tibble(year = rep(2001:2003, each =3),
       ID = c("A1", "A2", "A3",
              "A1", "A1", "A3",
              "A1", "A2", "A3"))
df
```

    #> # A tibble: 9 × 2
    #>    year ID   
    #>   <int> <chr>
    #> 1  2001 A1   
    #> 2  2001 A2   
    #> 3  2001 A3   
    #> 4  2002 A1   
    #> 5  2002 A1   
    #> 6  2002 A3   
    #> 7  2003 A1   
    #> 8  2003 A2   
    #> 9  2003 A3

``` r
#row 5 should fail test

library(pointblank)

#I think this worked in a previous version, but doesn't anymore?
create_agent(df) %>% 
  col_vals_lte(vars(n), 1,
               preconditions = ~ . %>%
                 group_by(ID, year) %>%
                 count()) %>% 
  interrogate()
```

    #> Warning in agent$validation_set$n[idx] <- row_count: number of items to replace
    #> is not a multiple of replacement length

    #> Warning in agent$validation_set$n_passed[idx] <- n_passed: number of items to
    #> replace is not a multiple of replacement length

    #> Warning in n_passed/row_count: longer object length is not a multiple of shorter
    #> object length

    #> Warning in agent$validation_set$f_passed[idx] <- round((n_passed/row_count), :
    #> number of items to replace is not a multiple of replacement length

    #> Warning in agent$validation_set$f_failed[idx] <- round((n_failed/row_count), :
    #> number of items to replace is not a multiple of replacement length

**Comment by rich-iannone** [comment]
Hey Eric, try using the relatively new segmentation feature. Here's an example:

```r
library(pointblank)
library(tidyverse)

df <- 
  tibble(
    year = rep(2001:2003, each = 3),
    ID = c("A1", "A2", "A3",
           "A1", "A1", "A3",
           "A1", "A2", "A3"
    )
  )

agent <-
  create_agent(
    tbl = df,
    actions = action_levels(warn_at = 1)
  ) %>%
  rows_distinct(segments = vars(year)) %>%
  interrogate()

agent
```

Here's a screen capture of the reporting:

<img width="1015" alt="pointblank-segmentation" src="https://user-images.githubusercontent.com/5612024/153272457-ce185d90-adf6-4438-8473-73db1a67062a.png">

**Reply by Aariq** [reply]
Ah, ok, this works except I'm not interested in entirely duplicated rows, just duplicates of one column.  For example, each ID number should only have one measurement of height per year.  If there are multiples, then someone entered the ID number wrong. Here's a slightly different example where row 5 should still fail.  

``` r
library(tidyverse)
#> Warning: package 'tidyr' was built under R version 4.0.5
df <- tibble(year = rep(2001:2003, each =3),
       ID = c("A1", "A2", "A3",
              "A1", "A1", "A3",
              "A1", "A2", "A3"),
       height = rnorm(9, 100)
       )
df
#> # A tibble: 9 × 3
#>    year ID    height
#>   <int> <chr>  <dbl>
#> 1  2001 A1     101. 
#> 2  2001 A2     102. 
#> 3  2001 A3     100. 
#> 4  2002 A1      99.3
#> 5  2002 A1     101. 
#> 6  2002 A3     100. 
#> 7  2003 A1     101. 
#> 8  2003 A2      99.2
#> 9  2003 A3      99.1
#row 5 should fail test

library(pointblank)
#> Warning: package 'pointblank' was built under R version 4.0.5

agent <-
  create_agent(
    tbl = df,
    actions = action_levels(warn_at = 1)
  ) %>%
  rows_distinct(segments = vars(year)) %>%
  interrogate()
```

<sup>Created on 2022-02-09 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>
<img width="983" alt="Screen Shot 2022-02-09 at 2 18 35 PM" src="https://user-images.githubusercontent.com/25404783/153274179-4940eb84-3e57-4d60-bf25-a38dec110475.png">

**Reply by Aariq** [reply]
something like a `col_vals_distinct()`

**Comment by rich-iannone** [comment, marked as answer]
Eric, I gotcha there too! With `rows_distinct()` you can focus on a subset of columns:

```r
library(pointblank)
library(tidyverse)

df <- 
  tibble(
    year = rep(2001:2003, each = 3),
    ID = c("A1", "A2", "A3",
           "A1", "A1", "A3",
           "A1", "A2", "A3"
    )
  )

agent <-
  create_agent(
    tbl = df,
    actions = action_levels(warn_at = 1)
  ) %>%
  rows_distinct(columns = "ID", segments = vars(year)) %>%
  interrogate()

agent
```

This yields this report:

<img width="1019" alt="pointblank-segmentation-2" src="https://user-images.githubusercontent.com/5612024/153274814-caa710a1-7d40-4ca4-9bec-0b06ca2aa49c.png">

**Reply by Aariq** [reply]
oh, thanks!  I missed that in the lengthy list of arguments.  Perfect!

**Reply by rich-iannone** [reply]
Awesome!
